### ISSUE TITLE
MessagesCollectionView bottom inset miscalculation when hardware keyboard is connected

### ISSUE THREAD

**Issue by austinwright** [MessageKit/MessageKit#690]
👻
## General Information
Issue calculating the MessagesCollectionView bottom inset when a hardware keyboard is attached to the device/simulator.

* MessageKit Version:
v 0.13.5 (though code looks exactly the same in the v1-beta-1)

* iOS Version(s):
iOS 11.2

* Swift Version:
Swift 4

* Devices/Simulators:
Simulator 7+/8+

* Reproducible in ChatExample? (Yes/No):
YES: (steps in simulator)
1) Make sure 'Simulator -> Hardware -> Keyboard -> Connect Hardware Keyboard' is checked
2) Focus the MessageInputBar
3) Type several newline characters
4.1) Delete one or more newline characters at once, OR
4.2) Tap Send
5) Observe

##  What happened?
The MessagesCollectionView inset is now too large and the collection view bottom is too far from the top of the MessageInputBar.

![screen shot 2018-05-16 at 1 48 56 pm](https://user-images.githubusercontent.com/1952159/40134291-f4eef8da-590f-11e8-9ce7-401895fc79a0.png)
![screen shot 2018-05-16 at 1 49 04 pm](https://user-images.githubusercontent.com/1952159/40134304-fee1c3e0-590f-11e8-96cb-bbbc3669a6b1.png)
![screen shot 2018-05-16 at 1 49 13 pm](https://user-images.githubusercontent.com/1952159/40134306-00734684-5910-11e8-8f18-651457fda750.png)

## What did you expect to happen?
The MessagesCollectionView inset is correctly calculated to match the top of the MessageInputBar

## Technical Info:
Bug location: MessagesViewController+Keyboard.handleKeyboardDidChangeState(_ notification: Notification)
Thoughts:
From what I observed, the first if statement's condition:
```swift
(keyboardEndFrame.origin.y + keyboardEndFrame.size.height) > UIScreen.main.bounds.height
```
returns false, because the two sides of the expression are equal (when the input bar grows they are equal too).

So the calculation moves onto the 'software keyboard' calculation, of which the key line is:
```swift
let afterBottomInset = keyboardEndFrame.height > keyboardOffsetFrame.height ? (keyboardEndFrame.height - iPhoneXBottomInset) : keyboardOffsetFrame.height
```
Since `afterBottomInset` eventually becomes the `MessagesCollectionView`'s bottom inset.
When the `MessageInputBar` grows, the `keyboardEndFrame.height` is greater than the `keyboardOffsetFrame.height` (aka. the old `MessageInputBar` height). So the bottom inset becomes the `keyboardEndFrame.height`.
When the `MessageInputBar` shrinks, however, the `keyboardEndFrame.height` is less than the `keyboardOffsetFrame.height` (aka. the old `MessageInputBar` height), and so the inset stays at the previous `MessageInputBar` height.

Maybe we could leverage the `MessageInputBar`'s `previousIntrinsicContentSize` and check if it is getting smaller to make more informed calculations? Or adjust the initial if condition to more accurately identify when we need to use the 'hardware' calculation.

(As a side note, if the 'hardware' `if` statement had been triggered, then I don't think the `maintainPositionOnKeyboardFrameChanged` setting would be honored).

**Comment by austinwright** [comment]
@nathantannar4, Looking at the blame history, it seems like you were involved in modifying these calculations most recently. If you take a look and LMK what you think and if you have any suggestions it would be much appreciated. (I can fix this, but I want to make sure it is fixed 'right').

**Comment by nathantannar4** [comment]
@austinwright Thank you for your detailed bug report, I will look into this further this weekend

**Comment by zhongwuzw** [comment]
Another issue about bottom inset, see #562 .

**Comment by zhongwuzw** [comment]
Fixed in #707 , @austinwright please let me know if any problem. Thanks for your issue.
